### PR TITLE
Implement Multiplatform support

### LINKED ISSUE #2: Migrate to kotlin multiplatform
PRs welcome!

### PR DESCRIPTION
**Pull request by X1nto**
# Additions
Supports the following platforms:
- [x] JVM
- [x] Android
- [x] iOS (arm64, arm64sim, x86_64)
- [x] macOS (arm64, x86_64)
- [x] Linux (arm64, x86_64)
- [x] Windows (MinGW x86_64) 
- [x] JS

~~Tests are currently failing on JS and I have no idea why, will investigate. They're successful on macOS and JVM~~

# Changes
- Refactors the library to use [the bignum library](https://github.com/ionspin/kotlin-multiplatform-bignum/)
	- Stores the internal `bytes` variable as BigDecimal
	- Exposes the BigDecimal, BigInteger, and other bignum APIs to the end user 
- Changes the names of the test functions, because Kotlin JS doesn't like spaces in names
- ~~Refactors the math functions to be platform-specific~~
- ~~Migrates to kotlin.test for multiplatform tests~~
- ~~Replaces the IllegalStateException with a custom FileSizePrecisionException (not really sure about the name, suggestions are more than welcome)~~
- ~~Downgrades Java to version 8, to support more Java/Android apps~~

# Caveats
~~Due to the fact that JS likes to be uniquely stupid, Kotlin doesn't support the `is` operators for any type of number except for Long, [since everything but Long is a JS Number ](https://kotlinlang.org/docs/js-to-kotlin-interop.html#kotlin-types-in-javascript). This results in the library relying on a `nonJs` source set, which separates JS decimal checking from normal Double checks.~~

---------
Closes #2

### CODE REVIEW

**Comment by X1nto** [comment]
I'm not sure that Linux and Windows support is possible without pulling in [TomMath](https://github.com/libtom/libtommath) or something like that, which I don't really know how to bind to Kotlin properly

File: build.gradle.kts [hunk 1]
```diff
@@ -12,7 +13,7 @@ allprojects {
   plugins.withType<JavaBasePlugin>().configureEach {
     extensions.configure<JavaPluginExtension> {
       toolchain {
-        languageVersion = JavaLanguageVersion.of(11)
+        languageVersion = JavaLanguageVersion.of(8)
```
Comment (saket) [inline build.gradle.kts]: Can you move this change to a separate PR? Including unrelated changes makes it difficult to review a PR.

File: gradle/libs.versions.toml [hunk 2]
```diff
@@ -2,17 +2,17 @@
...
 mavenPublish = "0.25.3" # https://github.com/vanniktech/gradle-maven-publish-plugin/releases
 dokka = "1.9.0" # https://github.com/kotlin/dokka/releases
-junit = "4.13.2"
 poko = "0.15.1" # https://github.com/drewhamilton/poko/releases
 metalava = "0.3.5"  # https://github.com/tylerbwong/metalava-gradle/releases
 
 [plugins]
 kotlin-jvm = { id = "org.jetbrains.kotlin.jvm", version.ref = "kotlin" }
+kotlin-multiplatform = { id = "org.jetbrains.kotlin.multiplatform", version.ref = "kotlin" }
+kotlin-test = { id = "org.jetbrains.kotlin" }
```
Comment (saket) [inline gradle/libs.versions.toml:12]: was this accidentally leftover?

File: library/src/commonTest/kotlin/me/saket/filesize/FileSizeTest.kt [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import kotlin.test.assertFailsWith
+import me.saket.filesize.FileSize.Companion.bytes
+import me.saket.filesize.FileSize.Companion.gigabytes
+import me.saket.filesize.FileSize.Companion.kilobytes
+import me.saket.filesize.FileSize.Companion.megabytes
+
+class FileSizeTest {
+
+  @Test
+  fun unitConversions() {
```
Comment (saket) [inline library/src/commonTest/kotlin/me/saket/filesize/FileSizeTest.kt]: let's use underscores for making these test functions readable?

File: library/src/commonTest/kotlin/me/saket/filesize/FileSizeTest.kt [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    assertEquals(expected = 2, actual = 2_000.bytes.inWholeKilobytes)
```
Comment (saket) [inline library/src/commonTest/kotlin/me/saket/filesize/FileSizeTest.kt]: these assertion APIs from `kotlin.test` are terrible. what was the reason for moving away from assertk which is already multiplatform compatible?

File: library/src/commonMain/kotlin/me/saket/filesize/FileSize.kt [hunk 5]
```diff
@@ -97,10 +81,10 @@ class FileSize(private val bytes: Long) : Comparable<FileSize> {
     @get:JvmName("bytes")
     inline val Number.bytes: FileSize
       get() {
-        if (this is Double) {
+        if (this.isDecimal()) {
           // Double#bytes is already a compilation error. This runtime error prevents
           // developers from casting doubles as Numbers to bypass the compilation error.
-          error(PrecisionLossErrorMessage)
+          throw FileSizePrecisionException()
```
Comment (saket) [inline library/src/commonMain/kotlin/me/saket/filesize/FileSize.kt]: Hmm I'm not convinced about the name either. Let's continue using `error()` for now?

File: library/src/jsMain/kotlin/me/saket/filesize/internals.js.kt [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  return BigDecimal(this.toString())
+    .divide(BigDecimal(other))
+    .toLong()
+}
+
+
+@JsModule("js-big-decimal")
+@JsNonModule
+@JsName("bigDecimal")
+private external class BigDecimal(value: dynamic) {
```
Comment (saket) [inline library/src/jsMain/kotlin/me/saket/filesize/internals.js.kt]: Let's switch the storage of bytes from `Long` to `BigDecimal` in `FileSize`. I'm thinking that'll simplify a lot of code in this PR:

* Most of these `expect/actual` functions for math calculations will get moved to `commonMain`.
* This external class wouldn't need to exist
* We'd be able to add support for linux and windows.

https://github.com/ionspin/kotlin-multiplatform-bignum/

**Review by saket — CHANGES_REQUESTED** [review]
thanks @X1nto!

File: build.gradle.kts (hunk 1 above)
Comment (X1nto) [reply build.gradle.kts]: Sure, no problem

File: library/src/commonTest/kotlin/me/saket/filesize/FileSizeTest.kt (hunk 4 above)
Comment (X1nto) [reply library/src/commonTest/kotlin/me/saket/filesize/FileSizeTest.kt]: I converted them just to please some of the temporary source sets I made while migrating, I didn't think converting them back to assertk would be needed but sure, I can fix that

File: library/src/jsMain/kotlin/me/saket/filesize/internals.js.kt (hunk 6 above)
Comment (X1nto) [reply library/src/jsMain/kotlin/me/saket/filesize/internals.js.kt]: In this case, we can probably move all of the calculations to BigDecimal and most likely get rid of issues with Double/Float precision, should we explore that possibility?

Comment (saket) [reply library/src/jsMain/kotlin/me/saket/filesize/internals.js.kt]: yep yep yep!

File: library/src/commonMain/kotlin/me/saket/filesize/FileSize.kt (hunk 5 above)
Comment (X1nto) [reply library/src/commonMain/kotlin/me/saket/filesize/FileSize.kt]: No longer throwing anything, since the library can now handle doubles and floats

**Comment by X1nto** [comment]
One thing concerning me now is the fact that the `throw_if_calculation_of_a_large_filesize_results_in_an_overflow` test is no longer necessary since the library now stores the bytes as BigDecimal, making it impossible to throw when FileSize with large numbers are constructed. Is this how we want to move forward? I'm looking forward to your input @saket since I'm not really sure if I'm doing that part correctly

**Comment by X1nto** [comment]
Actually, the exception *will* be thrown as soon as the user calls any of the `FileSize#inWholeXxx` properties, as per `BigDecimal#longValue(exactValue = true)`, so there's that

**Review by saket — APPROVED** [review]
Kotlin is complaining about compilation errors, but I can fix them in a follow-up. Thanks a bunch @X1nto!

**Comment by X1nto** [comment]
> Kotlin is complaining about compilation errors, but I can fix them in a follow-up. Thanks a bunch @X1nto!

Not gonna lie, haven't had any compilation errors but sometimes Intellij/gradle likes to skip over new errors if they appear deep in the development phase. You're welcome, love this project!

**Comment by saket** [comment]
Were you able to resolve `@JvmName` and friends in your IDE despite the project being multiplatform?

**Comment by saket** [comment]
This says I should be able to but I wonder what's missing: https://github.com/JetBrains/kotlin/blob/f45d92eebcfe3ac6c9b7f48c570f08eac9fce255/libraries/stdlib/common/src/kotlin/JvmAnnotationsH.kt#L43

**Comment by X1nto** [comment]
> Were you able to resolve `@JvmName` and friends in your IDE despite the project being multiplatform?

Yes, those annotations are under the kotlin.jvm package and can be imported just like anything else. The only difference is that in a non-multiplatform environment, those annotations are imported implicitly.

**Pull request merged**